common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -333,6 +345,146 @@ public StyledText prependPart(StyledTextPart part) {
+
+            for (String stringPart : stringParts) {
+                splitParts.add(new StyledText(
+                        List.of(new StyledTextPart(
+                                stringPart, part.getPartStyle().getStyle(), null, Style.EMPTY)),
+                        temporaryWorkaround,
+                        clickEvents,
+                        hoverEvents));
+            }
+        }
+
+        return splitParts.toArray(new StyledText[0]);
+    }
+
+    public StyledText substring(int beginIndex) {
+        return substring(beginIndex, length());
+    }
+
+    /**
+     * Returns a new {@link StyledText} that is a substring of this {@link StyledText}.
+     * <p> Note that {@link PartStyle.StyleType.NONE} is used for calculating the index.
+     * @param beginIndex the beginning index, inclusive
+     * @param endIndex the ending index, exclusive
+     * @return the new {@link StyledText}
+     */
+    public StyledText substring(int beginIndex, int endIndex) {
+        if (endIndex < beginIndex) {
+            throw new IndexOutOfBoundsException("endIndex must be greater than beginIndex");
+        }
+        if (beginIndex < 0) {
+            throw new IndexOutOfBoundsException("beginIndex must be greater than or equal to 0");
+        }
+        if (endIndex > length()) {
+            throw new IndexOutOfBoundsException("endIndex must be less than or equal to length()");
+        }
+
+        List<StyledTextPart> includedParts = new ArrayList<>();
+
+        int currentIndex = 0;
+
+        for (StyledTextPart part : parts) {
+            if (currentIndex >= beginIndex && currentIndex + part.length() < endIndex) {
+                // 1. This full part is included
+
+                includedParts.add(part);
+            } else if (currentIndex + part.length() >= beginIndex || currentIndex + part.length() > endIndex) {
+                // 2. This part is partially included
+
+                int startIndexInPart = Math.max(0, beginIndex - currentIndex);
+                int endIndexInPart = Math.min(part.length(), endIndex - currentIndex);
+
+                String includedSubstring =
+                        part.getString(null, PartStyle.StyleType.NONE).substring(startIndexInPart, endIndexInPart);
+
+                includedParts.add(new StyledTextPart(
+                        includedSubstring, part.getPartStyle().getStyle(), null, Style.EMPTY));
+            }
+
+            currentIndex += part.length();
+        }
+
+        return new StyledText(includedParts, clickEvents, hoverEvents);
+    }
+
+    /**
+     * Replaces the first occurrence of the given regex with the given replacement.
+     * <p> Note that {@link PartStyle.StyleType.NONE} is used when replacing.
+     * @param regex the regex to replace
+     * @param replacement the replacement
+     * @return the new {@link StyledText}
+     */
+    public StyledText replace(String regex, String replacement) {
```

Comment:
you should clarify that it does not match "foo§0bar" with regex foobar

Reply:
It is already mostly clarified in line 427